### PR TITLE
Restart Beat when certificates or key files are modified

### LINKED ISSUE #34408: Restart Beat on output cert/key file change
This issue derives from https://github.com/elastic/beats/issues/15956 and solves it for any Beat running on a Unix-like system that supports reexec.

Unfortunately properly reloading certificates and key files for an output requires either modifying the HTTPS connections (which requires modifying/replacing a some anonymous functions holding the TLS configuration) or fully reloading the outputs. Currently standalone Beats do not support output reload and when running under Elastic-Agent the default behaviour is to re-start the Beat on Output reload.

A quick solution to implement the certificate reload to mimic the "output reload" under Elastic-Agent by restarting the Beat. On Unix-like systems it can be done using reexec.

Partially solves: https://github.com/elastic/beats/issues/15956

### PR DESCRIPTION
**Pull request by belimawr**
## What does this PR do?

This PR enables any Beat to reexec when a output CA certificate file is modified. The main use case we're covering here is enabling "CA certificates reload" when running Metricbeat on ECK to monitor Elasticsearch. The "Related issues" section contains links with more details.

`libbeat/management/management.go` is also refactored and some unused bits of code are removed.

## Why is it important?

It enables Beats to re-start whenever the CA certificate for an output changes. This enables the Metricbeat side car used by ECK to monitor Elasticsearch to "reload" the certificates without having to restart Elasticsearch container.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Proper names for the loggers
- [x] Should we rename the `nilManager`? It has never been `nil`, and now it has got some function.
- [x] Fix config keys and docs
- [x] Finish Windows implementation
- [x] Add documentation and logs about the supported OSes
- [x] Document about the needed seccomp rule to allow the reexec syscall
- [x] Test on MacOS
- [x] Update PR description with restart instead of exit, etc.

## How to test this PR locally

1. Run any Beat with an elasticsearch output `ssl` section like this:
    ```yaml
      ssl:
        certificate_authorities: ["/path/to/your/ca/cert.pem"]
        verification_mode: full
        restart_on_cert_change:
          enabled: true
          period: 1m
    ```
3. Wait for the Beat to start
4. Touch one of the files defined on `output.elasticsearch.ssl.certificate_authorities`
5. The Beat restart. You can see the shutdown/start logs

You can also run the test defined on my branch: https://github.com/belimawr/beats/tree/reload-certificates-key-files-tests-2. Because of some global state, only one test from `beat_integration_test.go` can be run, hence it is not part of this PR.

1. Go to `beats/libbeat/cmd/instance`
2. Run the test with `go test -v -run="TestExit" -timeout=40s`
3. The test should pass.

## Related issues
- Relates #15956
- Closes #34408

## Use cases

- Reloading certificates on ECK

~~## Screenshots~~
~~## Logs~~

### CODE REVIEW

**Comment by rdner** [comment]
It's not clear to me why this should be implemented in Beats.

> It enables Beats to re-start (when running under a supervisor)

I was under impression that Elastic Agent generates/manages the certificates when configuring and running underlying processes. I would expect the agent to monitor the certificates and restart processes if the certificates expire. Am I missing some context here? Even if the certificates are not managed by the agent and just configured by the agent I still think these restarts have to be centralised in the agent.

Doing it from the Beats themselves, from my point of view, creates the following issues:

* Other input implementations (outside Beats) will have to implement the same (e.g. Endpoint), otherwise the behaviour will be inconsistent
* Beats will have to report this certificate modification to Elastic Agent in a special way before exiting, so the Fleet Server displays the right status and makes it transparent to the user why the process restarted
* What happens if the certificate file is deleted for whatever reason? It looks like we're going into the crash loop with the current implementation.

**Comment by belimawr** [comment]
> Am I missing some context here?

Yes, you are. I'm really sorry for that, I should have been more clear in the PR description. I'll update it soon.

That is not for Beats running under Elastic-Agent, it's for standalone Beats, this PR aims to solve a subset of the [original issue](https://github.com/elastic/beats/issues/15956) that is enabling "CA certificate reload" used by outputs and inputs and unblocking ECK. The idea is to re-start the Beats (Metricbeat on their case) without having to re-start the whole pod (an ES pod with Metricbeat as a side car).

On ECK, whenever Metricbeats exit, its container is re-started by Kubernetes, but ES one is left un-touched.

**Comment by rdner** [comment]
@belimawr I guess my misunderstanding came from the word "supervisor", I thought you meant Elastic Agent :)

File: libbeat/cmd/instance/beat.go [hunk 1]
```diff
@@ -984,6 +985,64 @@ func (b *Beat) createOutput(stats outputs.Observer, cfg config.Namespace) (outpu
...
 
+	// TODO (Tiago): find a proper name for this logger
+	logger := logp.L().Named("ssl.cert.reloader")
+	// Here the output is created and we have acces to the Beat struct (with the manager)
+	// as a workaround we can unpack the new settings and trigger the reload-watcher from here
+
+	// We get an output config, so we extract the 'SSL' bit from it
+	rawTLSCfg, err := cfg.Config().Child("ssl", -1)
+	if err != nil {
+		return outputs.Group{}, fmt.Errorf("could not extract the 'ssl' section of the output config: %w", err)
```
Comment (leehinman) [inline libbeat/cmd/instance/beat.go:996]: do we get an error if the config doesn't have an ssl section?, just want to make sure we can still get an output even if it doesn't have ssl.

File: libbeat/cmd/instance/beat.go [hunk 2]
```diff
@@ -984,6 +985,64 @@ func (b *Beat) createOutput(stats outputs.Observer, cfg config.Namespace) (outpu
...
+
+	watchers := make([]*cfgfile.GlobWatcher, 0, len(tlsCfg.CAs))
+	for _, ca := range tlsCfg.CAs {
+		if tlscommon.IsPEMString(ca) {
+			// That's an embedded cert, we're only interested in files
+			continue
+		}
+
+		logger.Debugf("watching '%s' for changes", ca)
+		gw := cfgfile.NewGlobWatcher(ca)
```
Comment (leehinman) [inline libbeat/cmd/instance/beat.go:1013]: this may or may not be an issue.  But the CA from the config point of view is a filepath, not a glob.  So if the filepath has something *, ?, [, ], or \\ in it we will get "interesting" results.

File: libbeat/cmd/instance/beat.go [hunk 3]
```diff
@@ -984,6 +985,64 @@ func (b *Beat) createOutput(stats outputs.Observer, cfg config.Namespace) (outpu
...
+			<-ticker
+			for _, gw := range watchers {
+				files, changed, err := gw.Scan()
+				if err != nil {
+					logger.Warnf("could not scan certificate files: %s", err.Error())
+				}
+
+				if changed {
+					logger.Infof(
+						"some of the following files have been modified: %v, starting %s shurdown",
```
Comment (leehinman) [inline libbeat/cmd/instance/beat.go:1038]:
```suggestion
						"some of the following files have been modified: %v, starting %s shutdown",
```

File: libbeat/cmd/instance/beat.go [hunk 4]
```diff
@@ -984,6 +985,64 @@ func (b *Beat) createOutput(stats outputs.Observer, cfg config.Namespace) (outpu
...
+						files, b.Info.Beat)
+					b.Manager.Stop()
```
Comment (leehinman) [inline libbeat/cmd/instance/beat.go:1040]: I think we want this behavior behind some kind of flag, or at least a warning of some kind.  As a system admin I wouldn't expect modifying a ca file to halt an application.  It would be fairly normal to pre-provision the files and schedule a restart for later.

File: libbeat/cmd/instance/beat.go (hunk 1 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:996]: That's one of the many parts missing, aside the happy path, I haven't really handled anything else.

File: libbeat/cmd/instance/beat.go (hunk 2 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:1013]: That's a good one, thanks!. I was wondering if I could use the filepath as a glob. On a quick look I did not find a "FileWatcher", so I went with the Glob one to prove the concept/get an idea of how the code will look like. I'll make a note to look into this later.

I wonder who would use any of those characters on a filename on a server... However we do need to cover it.

File: libbeat/cmd/instance/beat.go (hunk 4 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:1040]: Yes, definitely. I'll implement the configuration part soon.

For now I was looking more for feedback about the overall approach. I'm not a fan of parsing the config in different places, but that was the only way I found to get it working without a huge refactoring.

File: libbeat/cmd/instance/beat.go (hunk 2 above)
Comment (leehinman) [reply libbeat/cmd/instance/beat.go:1029]: I know it is an edge case, but then who would name their kid ["Robert'); DROP Table Students;--"](https://xkcd.com/327/) ? 😃

File: libbeat/cmd/instance/beat.go (hunk 4 above)
Comment (leehinman) [reply libbeat/cmd/instance/beat.go:1055]: I'm ok with the approach given how we specify CAs today.  It might be nice to change how we specify CAs, so you can list a directory and then we read in all the certs from that directory.  That way you don't have to use the same name as listed in the current config file.

Comment (belimawr) [reply libbeat/cmd/instance/beat.go:1055]: That would be interesting, but might as well be a breaking change :/

To be honest, I don't know what is more common out there: to use a folder for each application or have explicitly paths set.

Because its a central piece of security, I like being more explicit, but it does bring other "issues".

File: libbeat/cmd/instance/beat.go (hunk 2 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:1029]: I love his one, hauahuahuahua

But for some reason I always thought his name was Bob. Or maybe I'm just too used  to Alice, Bob and Eve :P

File: libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl [hunk 5]
```diff
@@ -81,6 +81,14 @@ output.elasticsearch:
   #allow_older_versions: false
 
 {{include "ssl.reference.yml.tmpl" . | indent 2 }}
+
+  # Enable exiting {{.BeatName}} if any of the files pointed by
+  # `ssl.certificate_authorities` changes
+  #ssl.exit_on_cert_change.enabled
+
+  # Interval to scan for changes on CA certificate files
+  #ssl.exit_on_cert_change.period
```
Comment (belimawr) [inline libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl:90]: Now the real hard question: Is the `ssl` a good place for this config? I thought about putting on the output config or even as a general beat config parameter, but I settled with `ssl` because it's specific for those certs.

**Comment by belimawr** [comment]
This PR is now ready for review. Currently it is using my fork of `elastic-agent-libs` so I can have the final version of the code, but once https://github.com/elastic/elastic-agent-libs/pull/97, I'll update `go.mod`.

There are no tests because the tests "start the Beat" without spanning a new process, break if there is more than one on the package :man_facepalming:. I move the tests to another branch on my fork: https://github.com/belimawr/beats/tree/reload-certificates-key-files-tests, if someone is interested, the test can be successfully run with:
```shell
cd libbeat/cmd/instance
go test -v -run="TestExit" -timeout=30s               
```
I'm planning to add the tests on another PR.

File: libbeat/cmd/instance/beat.go [hunk 6]
```diff
@@ -979,11 +982,104 @@ func (b *Beat) makeOutputFactory(
...
+func (b *Beat) reloadOutputOnCertChange(cfg config.Namespace) error {
+	// TODO (Tiago): find a proper name for this logger
+	logger := logp.L().Named("ssl.cert.reloader")
+	// Here the output is created and we have acces to the Beat struct (with the manager)
+	// as a workaround we can unpack the new settings and trigger the reload-watcher from here
+
+	// We get an output config, so we extract the 'SSL' bit from it
+	rawTLSCfg, err := cfg.Config().Child("ssl", -1)
+	if err != nil {
+		e, ok := err.(ucfg.Error)
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:994]: I think it's worth to fix the linter issue here (avoid the type casting).

File: libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl [hunk 7]
```diff
@@ -81,6 +81,14 @@ output.elasticsearch:
   #allow_older_versions: false
 
 {{include "ssl.reference.yml.tmpl" . | indent 2 }}
+
+  # Enable exiting {{.BeatName}} if any of the files pointed by
+  # `ssl.certificate_authorities` changes
+  #ssl.exit_on_cert_change.enabled: false
+
+  # Interval to scan for changes on CA certificate files
+  #ssl.exit_on_cert_change.period: 1m
```
Comment (rdner) [inline libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl:90]: Nitpicking but why different words are used? `interval` vs `period`? I'd rather use `interval` everywhere.

File: libbeat/cmd/instance/beat.go [hunk 8]
```diff
@@ -979,11 +982,104 @@ func (b *Beat) makeOutputFactory(
...
+		logger.Debugf("watching '%s' for changes", f)
+		filesToWatch = append(filesToWatch, f)
+	}
+
+	watcher := filewatcher.New(filesToWatch...)
+	// Ignore the first scan as it will always return
+	// true for files changed. The output has not been
+	// started yet, so even if the files have changed since
+	// the Beat started, they don't need to be reloaded
+	watcher.Scan()
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:1049]:
```suggestion
	_ = watcher.Scan()
```

File: libbeat/cmd/instance/beat.go [hunk 9]
```diff
@@ -979,11 +982,104 @@ func (b *Beat) makeOutputFactory(
...
+
+	// Watch for file changes while the Beat is alive
+	go func() {
+		ticker := time.Tick(extendedTLSCfg.Reload.Period)
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:1053]:
```suggestion
		ticker := time.NewTicker(extendedTLSCfg.Reload.Period)
		defer ticker.Stop()
```

This will fix the linter.

File: libbeat/cmd/instance/beat.go [hunk 10]
```diff
@@ -979,11 +982,104 @@ func (b *Beat) makeOutputFactory(
...
+
+		for {
+			<-ticker
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:1056]:
```suggestion
			<-ticker.C
```

File: libbeat/cmd/instance/beat.go [hunk 11]
```diff
@@ -979,11 +982,104 @@ func (b *Beat) makeOutputFactory(
...
+	}{
+		Reload: cfgfile.Reload{
+			Enabled: false,
+			Period:  time.Minute,
+		},
+	}
+
+	if err := rawTLSCfg.Unpack(&extendedTLSCfg); err != nil {
+		return fmt.Errorf("unpacking 'ssl' config: %w", err)
+	}
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:1017]: Why there is no validation after this? We should validate the interval to be greater or equal to 1 second and the docs should not recommend this limit but explicitly say it's enforced by the validation.

File: libbeat/docs/shared-ssl-config.asciidoc [hunk 12]
```diff
@@ -567,6 +567,23 @@ Allows a remote server to request renegotiation repeatedly.
...
+NOTE: This feature is intended to be used when {beatname_uc} is running
+under a supervisor that can restart it. The default value is `false`.
+
+[float]
+[[exit_on_cert_change_period]]
+==== `exit_on_cert_change.period`
+Specifies how often the files are checked for changes. Do not set the
+period to less than 1s because the modification time of files is often
+stored in seconds. Setting the period to less than 1s will result in
+unnecessary overhead. The default value is 1m.
```
Comment (rdner) [inline libbeat/docs/shared-ssl-config.asciidoc:585]: This should be enforced by validation, fewer SDHs in the future.

File: libbeat/management/management.go [hunk 13]
```diff
@@ -152,30 +152,26 @@ func defaultModeConfig() *modeConfig {
 	}
 }
 
-// nilManager, fallback when no manager is present
-type nilManager struct {
+// fallbackManager, fallback when no manager is present
```
Comment (rdner) [inline libbeat/management/management.go:155]: Looks like this file has a bunch of unused types and variables we can remove (see above).

File: libbeat/cmd/instance/beat.go [hunk 14]
```diff
@@ -979,11 +982,104 @@ func (b *Beat) makeOutputFactory(
...
+
+	extendedTLSCfg := struct {
+		tlscommon.Config `config:",inline" yaml:",inline"`
+		Reload           cfgfile.Reload `config:"exit_on_cert_change" yaml:"exit_on_cert_change"`
+	}{
+		Reload: cfgfile.Reload{
+			Enabled: false,
+			Period:  time.Minute,
+		},
+	}
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:1013]: I think it would be better to define it as a separate type.
Should we also define a `default` function for the `Reload` part in the `cfgfile` package?

It's kinda difficult to find these default values inside the function body here, if you need to, especially for a new comer.

File: libbeat/cmd/instance/beat.go [hunk 15]
```diff
@@ -979,11 +982,104 @@ func (b *Beat) makeOutputFactory(
 	}
 }
 
+func (b *Beat) reloadOutputOnCertChange(cfg config.Namespace) error {
+	// TODO (Tiago): find a proper name for this logger
```
[the remark this answers is not in the document]
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:986]: It kinda still is. So far nobody mentioned anything regarding the logger name. I'll assume it's good enough and remove this comment.

File: libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl (hunk 7 above)
Comment (belimawr) [reply libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl:90]: I use `period` as config key to be consistent with config reload (I'm using the same struct):
https://github.com/elastic/beats/blob/ad651a79276c3570f6077723f5df02bf3ea2da36/filebeat/filebeat.reference.yml#L1249-L1250

But I can update the comment to use "period" instead of "interval".

File: libbeat/cmd/instance/beat.go (hunk 8 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:1049]: `Scan` returns 3 values, it would have to be:
```
	_, _, _ = watcher.Scan()
```

No strong opinion here, I can change it.

File: libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl (hunk 7 above)
Comment (rdner) [reply libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl:90]: I'm for consistency. In the config docs it's not using the word `interval` though https://www.elastic.co/guide/en/beats/filebeat/current/_live_reloading.html

Double-wording is a bit confusing, it can cause someone using `interval` as the config key instead of `period`.

It's not a strong opinion though.

File: libbeat/cmd/instance/beat.go (hunk 8 above)
Comment (rdner) [reply libbeat/cmd/instance/beat.go:1049]: Well, we need this to fix the linter issue.

File: libbeat/cmd/instance/beat.go (hunk 9 above)
Comment (rdner) [reply libbeat/cmd/instance/beat.go:1053]: But why? There is no benefit from using `time.Tick` over `time.NewTicker`, it's worth to change to fix the linter, it's the same effort as adding the `nolint` directive.

File: libbeat/management/management.go (hunk 13 above)
Comment (belimawr) [reply libbeat/management/management.go:155]: I'll clean them up!

File: libbeat/cmd/instance/beat.go (hunk 14 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:1013]: > I think it would be better to define it as a separate type.

Agreed, I'll move it somewhere else on this file

> Should we also define a default function for the Reload part in the cfgfile package?

Different parts of Beats use this `Reload` struct, I believe the default value for it depends on the feature rather on the struct itself.

File: libbeat/cmd/instance/beat.go (hunk 9 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:1053]: Using `defer` on a function that never ends has no effect. Essentially it's dead code. It might even lead the reader to believe the deferred code will eventually run.

That is the indented purpose of `time.Tick`, from the docs:
```
// Tick is a convenience wrapper for NewTicker providing access to the ticking
// channel only. While Tick is useful for clients that have no need to shut down
// the Ticker (...)
```

If the point here is to stop the linter errors, the comment fixes the linter without adding dead code.

File: libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl (hunk 7 above)
Comment (belimawr) [reply libbeat/_meta/config/output-elasticsearch.reference.yml.tmpl:90]: I updated the comment to use "period", I tried to keep it short here because it's the reference doc. I'll push my changes very soon.

File: libbeat/cmd/instance/beat.go (hunk 9 above)
Comment (rdner) [reply libbeat/cmd/instance/beat.go:1053]: If you think it's better to write a `nolint` directive instead of a `defer` statement, go ahead, it would not block the review. Either way we need to fix the linter here.

File: libbeat/cmd/instance/beat.go (hunk 14 above)
Comment (rdner) [reply libbeat/cmd/instance/beat.go:1013]: > Different parts of Beats use this Reload struct, I believe the default value for it depends on the feature rather on the struct itself.

I think it makes sense to return the whole `extendedTLSCfg` from a `defaultExtendedTLSCfg` defined in this file then. It should be semantically marked as `default` somehow, I believe this is the way :)

File: libbeat/cmd/instance/beat.go [hunk 16]
```diff
@@ -138,6 +141,18 @@ type beatConfig struct {
...
+	tlscommon.Config `config:",inline" yaml:",inline"`
+	Reload           cfgfile.Reload `config:"exit_on_cert_change" yaml:"exit_on_cert_change"`
+}
+
+func (c certReloadConfig) Validate() error {
+	if c.Reload.Period < time.Second {
+		return errors.New("'exit_on_cert_change.period' must be equal or greather than 1s")
+	}
+	return nil
+}
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:154]: Is it called automatically by `.Unpack` because of some interface convention? I don't see it explicitly used anywhere.

Comment (belimawr) [reply libbeat/cmd/instance/beat.go:154]: Yes, it is.

**Comment by belimawr** [comment]
I've addressed the last comments and updated the code with the latest version of elastic-agent-libs. It's ready for final review now.

**Comment by belimawr** [comment]
I also improved the section about how to test.

**Comment by belimawr** [comment]
The tests are failing due to expired certificates, this PR will fix it: https://github.com/elastic/beats/pull/34500

Once it's merged to main, I'll update this branch.

File: libbeat/cmd/instance/beat.go [hunk 17]
```diff
@@ -979,11 +1003,95 @@ func (b *Beat) makeOutputFactory(
...
+		}
+		if tlscommon.IsPEMString(f) {
+			// That's an embedded cert, we're only interested in files
+			continue
+		}
+
+		logger.Debugf("watching '%s' for changes", f)
+		filesToWatch = append(filesToWatch, f)
+	}
+
```
Comment (leehinman) [inline libbeat/cmd/instance/beat.go:1054]: Minor. It looks like filesToWatch could be empty. If that is true I think we can just return nil. I know it won't cause problems, but it would mean that a go routine is started and we have to watch for a ticker to fire.

**Review by leehinman — APPROVED** [review]
One minor comment that can be addressed later if necessary.  Otherwise LGTM.

File: libbeat/cmd/instance/beat.go (hunk 17 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:1054]: Good catch! I fixed it on a015403dc3

**Review by rdner — APPROVED** [review]

**Comment by belimawr** [comment]
This PR broke the system tests on Filebeat. 

TL;DR
With the changes introduced by this PR, the `--once` flag does not work as the tests on `filebeat/tests/system/test_modules.py` expect, hence they all fail.

I'm currently trying to get a better understanding of the best way forward for this feature/PR.

File: libbeat/cmd/instance/beat.go [hunk 18]
```diff
@@ -139,6 +144,27 @@ type beatConfig struct {
...
 
+type certReloadConfig struct {
+	tlscommon.Config `config:",inline" yaml:",inline"`
+	Reload           cfgfile.Reload `config:"restart_on_cert_change" yaml:"restart_on_cert_change"`
+}
+
+func (c certReloadConfig) Validate() error {
+	if c.Reload.Period < time.Second {
+		return errors.New("'restart_on_cert_change.period' must be equal or greather than 1s")
+	}
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:155]: We should also add

```
if c.Reload.Enabled && runtime.GOOS == "windows" {
  return errors.New("'restart_on_cert_change' is not supported on Windows")
}

Comment (rdner) [inline libbeat/cmd/instance/beat.go:155]: We should also add the platform validation here.

```suggestion
	if c.Reload.Period < time.Second {
		return errors.New("'restart_on_cert_change.period' must be equal or greather than 1s")
	}
	if c.Reload.Enabled && runtime.GOOS == "windows" {
		return errors.New("'restart_on_cert_change' is not supported on Windows")
	}
```

Comment (rdner) [reply libbeat/cmd/instance/beat.go:155]: To accept the suggestion you need a new import `runtime`.

File: libbeat/cmd/instance/beat.go [hunk 19]
```diff
@@ -980,11 +1018,106 @@ func (b *Beat) makeOutputFactory(
 	}
 }
 
+func (b *Beat) reloadOutputOnCertChange(cfg config.Namespace) error {
+	logger := logp.L().Named("ssl.cert.reloader")
+	// Here the output is created and we have acces to the Beat struct (with the manager)
```
Comment (rdner) [inline libbeat/cmd/instance/beat.go:1023]:
```suggestion
	// Here the output is created and we have access to the Beat struct (with the manager)
```

File: libbeat/common/seccomp/policy_linux_arm.go [hunk 20]
```diff
@@ -26,7 +26,6 @@ func init() {
 			{
 				Action: seccomp.ActionErrno,
 				Names: []string{
-					"execve",
```
Comment (rdner) [inline libbeat/common/seccomp/policy_linux_arm.go:29]: What are these changes? Why do we remove it from `arm` and add to `amd64`?

Comment (belimawr) [reply libbeat/common/seccomp/policy_linux_arm.go:29]: ARM has got a block list for seccomp, so I remove `execve` from the block list. L27 states the action for those syscalls: `seccomp.ActionErrno`.

File: libbeat/cmd/instance/beat.go (hunk 18 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go:155]: I ended up adding both your changes on 562a476c37. Only when committing  the changes I realised `runtime` is already imported on this file.

Comment (belimawr) [reply libbeat/cmd/instance/beat.go:155]: Thanks a lot for already suggesting the validation code :heart:

File: libbeat/cmd/instance/beat.go (hunk 19 above)
Comment (belimawr) [reply libbeat/cmd/instance/beat.go]: Done on 562a476c37

**Review by rdner — APPROVED** [review]
Re-approving again, the current state LGTM.
